drivers/mrf24j40/mrf24j40_netdev.c

```diff
@@ -0,0 +1,578 @@
+                  (unsigned)len + 2);
+            return -EOVERFLOW;
+        }
+
+#ifdef MODULE_NETSTATS_L2
+        netdev->stats.tx_bytes += len;
+#endif
+        len = mrf24j40_tx_load(dev, ptr->iov_base, ptr->iov_len, len);
+        if (i == 0) {
+            dev->header_len = len;
+        }
+
+    }
+
+    /* send data out directly if pre-loading is disabled */
+    if (!(dev->netdev.flags & MRF24J40_OPT_PRELOADING)) {
+        mrf24j40_tx_exec(dev);
+    }
+    /* return the number of bytes that were actually send out */
+    return (int)len;
+
+}
+
+static int _recv(netdev2_t *netdev, void *buf, size_t len, void *info)
+{
+    mrf24j40_t *dev = (mrf24j40_t *)netdev;
+    uint8_t phr;
+    size_t pkt_len;
+
+    /* Turn off reception of packets off the air. This prevents the
+     * device from overwriting the buffer while we're reading it.
+     * BBREG1 only contains the RXDECINV setting, so no overwriting
+     * other bits with this.
+     */
+    mrf24j40_reg_write_short(dev, MRF24J40_REG_BBREG1, MRF24J40_BBREG1_RXDECINV );
+
+    /* get the size of the received packet */
+    phr = mrf24j40_reg_read_long(dev, MRF24J40_RX_FIFO);
+
+    pkt_len = (phr & 0x7f) - 2;
+
+    /* just return length when buf == NULL */
+    if (buf == NULL) {
+        return pkt_len;
+    }
+#ifdef MODULE_NETSTATS_L2
+    netdev->stats.rx_count++;
+    netdev->stats.rx_bytes += pkt_len;
+#endif
+    /* not enough space in buf */
+    if (pkt_len > len) {
+        DEBUG("[mrf24j40] No space in receive buffers\n");
+        mrf24j40_reg_write_short(dev, MRF24J40_REG_RXFLUSH, 0x01);
+        /* Turn on reception of packets off the air */
+        mrf24j40_reg_write_short(dev, MRF24J40_REG_BBREG1, 0x00);
+        return -ENOBUFS;
+    }
+    /* copy payload */
+    mrf24j40_rx_fifo_read(dev, 1, (uint8_t *)buf, pkt_len);
+
+    if (info != NULL) {
+        netdev2_ieee802154_rx_info_t *radio_info = info;
+        /* Read LQI and RSSI values from the RX fifo */
+        mrf24j40_rx_fifo_read(dev, phr + 1, &(radio_info->lqi), 1);
+        mrf24j40_rx_fifo_read(dev, phr + 2, &(radio_info->rssi), 1);
+    }
+
+    /* Turn on reception of packets off the air */
+    mrf24j40_reg_write_short(dev, MRF24J40_REG_BBREG1, 0x00);
+    return pkt_len;
+}
+
+static netopt_state_t _get_state(mrf24j40_t *dev)
+{
+    if (!(dev->pending & MRF24J40_TASK_TX_DONE)) {
+        return NETOPT_STATE_TX;
+    }
+    if (dev->pending & MRF24J40_TASK_RX_READY) {
+        return NETOPT_STATE_RX;
+    }
+    switch (dev->state) {
+        case MRF24J40_PSEUDO_STATE_SLEEP:
+            return NETOPT_STATE_SLEEP;
+    }
+    return NETOPT_STATE_IDLE;
+}
+
+static int _get(netdev2_t *netdev, netopt_t opt, void *val, size_t max_len)
+{
+    mrf24j40_t *dev = (mrf24j40_t *) netdev;
+
+    if (netdev == NULL) {
+        return -ENODEV;
+    }
+
+    int res;
+    switch (opt) {
+        case NETOPT_CHANNEL_PAGE:
+            if (max_len < sizeof(uint16_t)) {
```

Comment:
We should consider to replace this with `asserts` as done in #4419. But no worries. That can eb done in a follow up PR.